lib/ansible/module_utils/_text.py

```diff
@@ -276,3 +279,68 @@ def to_text(obj, encoding='utf-8', errors=None, nonstring='simplerepr'):
     to_native = to_text
 else:
     to_native = to_bytes
+
+
+def _wrap_dict(to_func, obj, **kwargs):
+    return dict(
+        (_wrap_to_string(to_func, k, **kwargs), _wrap_to_string(to_func, item, **kwargs))
+        for k, item in obj.items()
+    )
+
+
+def _wrap_sequence(to_func, obj, **kwargs):
+    obj_type = type(obj)
+    return obj_type(_wrap_to_string(to_func, item, **kwargs) for item in obj)
+
+
+def _wrap_set(to_func, obj, **kwargs):
+    return set(_wrap_to_string(to_func, item, **kwargs) for item in obj)
+
+
+def _wrap_to_string(to_func, obj, **kwargs):
+    r"""Ensure objects in a nested data structure are forced to
+    be a text/byte/native string depedending on given fuction:
+
+        :func:`ansible.module_utils._text.to_text` or
+        :func:`ansible.module_utils._text.to_bytes` or
+        :func:`ansible.module_utils._text.to_native`.
+
+    This is a private function and should not be used outside of the three
+    partial functions: wrap_to_text, wrap_to_bytes, wrap_to_native; which
+    are defined below in this file.
+
+    :arg to_func: The wrapping function to use to transform the object.
+    :arg obj: An object to make sure is wrapped by `to_func` function.
+    :kwarg \*\*kwargs: Optional arguments that function passed in ``to_func`` takes.
+
+    :returns: This returns given `obj` object and its nested objects wrapped in
+        `to_func` function.
+
+    .. version_added:: 2.10
+    """
+    if obj is None:
+        return obj
+
+    if isinstance(obj, (text_type, binary_type)):
+        obj = to_func(obj, **kwargs)
+    elif isinstance(obj, Mapping):
+        obj = _wrap_dict(to_func, obj, **kwargs)
+    elif isinstance(obj, Set):
+        obj = _wrap_set(to_func, obj, **kwargs)
+    elif is_sequence(obj):
+        obj = _wrap_sequence(to_func, obj, **kwargs)
+
+    return obj
+
+
+wrap_to_text = partial(_wrap_to_string, to_text)
+wrap_to_text.__doc__ = ("Ensure strings in a nested data structure are forced "
+                        "to be text strings.")
+
+wrap_to_bytes = partial(_wrap_to_string, to_bytes)
+wrap_to_bytes.__doc__ = ("Ensure strings in a nested data structure are forced "
+                         "to be byte strings.")
+
+wrap_to_native = partial(_wrap_to_string, to_native)
+wrap_to_native.__doc__ = ("Ensure strings in a nested data structure are forced "
+                          "to be native strings.")
```

Comment:
We do have code that does this already. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/text/converters.py#L44. That file is also where to_text/bytes/native should move eventually.

Your impregnation may be better (it looks like an appropriate use of partial, for instance).

One thing to note, though.  Your implementation will end up converting compatible types which may not be desirable in a generic function.  For instance, tuples would become lists.  Frozensets would become sets. NamedTuples would become a list.

I'd probably have this pr uses the code in common.text.converters and open a seperate pr to change the behaviour of container_to_* that you, samdoran, and i can discuss when he gets back.  (If you need set support now, that can probably be added, using the concrete set type, though)

Reply:
> We do have code that does this already. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/common/text/converters.py#L44. That file is also where to_text/bytes/native should move eventually.

Sigh, I missed that :-(

> 
> Your implementation may be better (it looks like an appropriate use of partial, for instance).
> 
> One thing to note, though. Your implementation will end up converting compatible types which may not be desirable in a generic function. For instance, tuples would become lists. Frozensets would become sets. NamedTuples would become a list.

This PR contains unit tests for `tuples`, they seem to be preserved:
```
>>> from ansible.module_utils._text import wrap_to_text
>>> a_tuple = (b'foo', b'bar')
>>> wrap_to_text(a_tuple)
('foo', 'bar')
>>> type(wrap_to_text(a_tuple))
<class 'tuple'>
```

`frozensets` are not, `namedtuples` raise an exception. I will look into this.

> 
> I'd probably have this pr use the code in common.text.converters and open a seperate pr to change the behaviour of container_to_* that you, samdoran, and i can discuss when he gets back. (If you need set support now, that can probably be added, using the concrete set type, though)

Sounds good to me. The sets support is not needed for `safe_eval` as I don't think it is expected to handle them.

Thank you for looking into this!